sql/mysql/driver.go

```diff
@@ -76,6 +79,65 @@ func (d *Driver) NormalizeSchema(ctx context.Context, s *schema.Schema) (*schema
 	return d.dev().NormalizeSchema(ctx, s)
 }
 
+// Lock implements the schema.Locker interface.
+func (d *Driver) Lock(ctx context.Context, name string, timeout time.Duration) (unlock func() error, err error) {
+	var (
+		closer io.Closer
+		conn   = d.conn.ExecQuerier
+	)
+	// If the connection is sql.DB (a connection-pool),
+	// acquire a connection and use it to obtain the lock.
+	if opener, ok := d.ExecQuerier.(interface {
+		Conn(context.Context) (*sql.Conn, error)
+	}); ok {
+		c, err := opener.Conn(ctx)
+		if err != nil {
+			return nil, err
+		}
+		defer func() {
+			if err != nil {
+				// Return the connection to the pool,
+				// if we failed to obtain the lock.
+				_ = closer.Close()
+			}
+		}()
+		conn, closer = c, c
+	}
+	rows, err := conn.QueryContext(ctx, "SELECT GET_LOCK(?, ?)", name, int(timeout.Seconds()))
+	if err != nil {
+		return nil, err
+	}
+	var acquired sql.NullBool
+	switch err := sqlx.ScanOne(rows, &acquired); {
+	case err != nil:
+		return nil, err
+	case !acquired.Valid:
+		// NULL is returned in case of an unexpected internal error.
+		return nil, fmt.Errorf("sql/mysql: unexpected internal error on Lock(%q, %s)", name, timeout)
+	case !acquired.Bool:
+		return nil, schema.ErrLockTimeout
+	}
+	// Acquired.
+	return func() error {
```

Comment:
we can consider

```go
LockReleaser interface {
Lock(ctx context.Context, name string, timeout time.Duration) error
Release(name string) error
}
```

Reply:
> I am not sure, if returing a releaser func is a god way to go. I in this case have to carry the function around, when I want to release the lock instead of doing so on the Driver.

The connection should be attached to the Lock. A schema can acquire more than one lock.